htlcswitch/switch.go

```diff
@@ -1170,6 +1191,47 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
 	}
 }
 
+// CheckCircularForward checks whether a forward is circular (arrives and
+// departs on the same link) and returns a link error if the switch is
+// configured to disallow this behaviour.
+func (s *Switch) CheckCircularForward(incoming lnwire.ShortChannelID,
+	outgoing ChannelLink, paymentHash lntypes.Hash) *LinkError {
+
+	// Check that the incoming link is different to the outgoing link.
+	// If they are equal, the htlc is part of a circular route which may
+	// be used to lock up our liquidity.
+	isCircularRoute := incoming.ToUint64() ==
+		outgoing.ShortChanID().ToUint64()
+
+	// If the route is not circular we do not need to perform any further
+	// checks.
+	if !isCircularRoute {
+		return nil
+	}
+
+	// If the switch is configured to allow circular routes, log that we
+	// are allowing the route then return nil.
+	if s.cfg.AllowCircularRoute {
+		log.Info("allowing circular route over link: %v "+
+			"(payment hash: %x)", paymentHash,
+			outgoing.ChannelPoint())
+		return nil
+	}
+
+	// If our node disallows circular routes, get a channel update for
+	// the outgoing link and return a temporary channel failure. If we
+	// fail getting the channel update, we return a temporary node failure.
+	update, err := outgoing.ChannelUpdate()
+	if err != nil {
+		return NewLinkError(&lnwire.FailTemporaryNodeFailure{})
+	}
+
+	return NewDetailedLinkError(
+		lnwire.NewTemporaryChannelFailure(update),
```

Comment:
I don't think we need to return an update. There isn't a problem with the policy.

Reply:
Nil update is fine because the spec has a length field, right? It doesn't specifically say whether we have to include an update. 

Also double checking, `TemporaryChannelUpdate` is the correct error to send?
Seems like the most fitting one in the bolts: "The channel from the processing node was unable to handle this HTLC, but may be able to handle it, or others, later."

Reply:
Yes, I think that is ok. It doesn't matter much which error we return, because the sender must know what they are doing.